# MAC-ErrorReads
MAC-ErrorReads is a machine learning-assisted classifier for filtering erroneous NGS reads. It converts the process of filtering erroneous NGS reads into a machine learning classification problem. MAC-ErrorReads learns a mapping function F that transforms the input feature space X extracted from each sequencing read into a binary label classification Y of this read as (1) for an erroneous read and (0) for a correct one.

# Usage
We are using MAC-ErrorReads to classify the sequencing reads as error-free or error redas. We first train our models using simulated datasets. We used the Wesim simulator to simulate our data, and after simulating the data, we split the data into training and testing.

We train different machine learning classifiers: Naive Bayes (NB), Support Vector Machine (SVM), Random Forest (RF), Logistic Regression (LR), and Extreme Gradient Boosting (XGBoost) to filter correct or erroneous reads. The models are trained on a labelled data set that has correct reads with label 0 and erroneous reads with label 1. We trained machine learning models using two reference genomes, and the training process could be generalised to different genomes with similar training steps.

To evaluate the efficacy of our trained models, we employed reads obtained from simulated data and real sequencing experiments that corresponded to the previously trained reference genomes.

The real reads in the sequencing experiments lacked explicit labels, and their accuracy levels were unknown beforehand.

To address this issue in the first dataset (E. coli), we proceeded with aligning the reads to their reference genome through the use of one of the NGS aligners (i.e., BWA). Subsequently, we extracted the computed alignment score for each read. By comparing the predicted labels generated by our trained machine learning models with the labels inferred from the alignment scores, we could effectively evaluate the performance of these models.

To address this issue in the second dataset (S. aureus), we assembled the reads generated after testing the data using Velvet assembler. After the assembly results are evaluated by one of the assembly evaluation tools called QUAST.We then used Lighter to correct the reads that were classified as errors and assembled the files that contained correct reads and errors that were corrected by Lighter.


# Tools
We are using different tools in our experment.

We used the wgsim (https://github.com/lh3/wgsim) simulator. Wgsim is a small tool for simulating sequence reads from a reference genome.

We also used the BWA aligner (https://github.com/lh3/bwa) to align the reads to the E. coli reference genome and extract their alignment scores from the resulting SAM files.

We also used Velvet assembler (https://github.com/dzerbino/velvet) to assemble the reads that are classified as error-free reads.

We also used QUAST (https://github.com/ablab/quast), the assembly evaluation tool, to evaluate the assembly results.

We are finally using Lighter (https://github.com/mourisl/Lighter), a KMER-based error correction method for genome sequencing data.We are using Lighter to correct the error reads.



# Data
We are using two different datasets. The first dataset used in our experiments is Escherichia coli str. K-12 substr. MG1655 (E.coli) with RefSeq accession entry NC_000913. The second dataset used in our experiments is the S. aureus genome from the GAUGE project.

We are using a simulated dataset to train our models. We used the wgsim simulator, which is included within the SAMtools for whole genome simulation.
To test our models, we are using simulated and real datasets.

The link to the simulated datasets used (https://drive.google.com/file/d/10nEAroKXB9uUEFjL8eZ6tle8wr5GryzX/view?usp=drive_link) to train and test our models

The first real dataset used to test our models is (E.coli) with accession number SRR625891, and the second real dataset used in our experiments is the S. aureus genome from the GAUGE project.

# License
This repository is under MIT license.

# Contact
Please do not hesitate to contact us (amira_samy@mans.edu.eg) if you have any comments, suggestions, or clarification requests regarding the study or if you would like to contribute to this resource.





